compiler/rustc_typeck/src/check/method/prelude2021.rs

```diff
@@ -206,6 +210,32 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             return;
         }
 
+        // For from_iter, check if the type actualy implements FromIterator.
+        // If we know it does not, we don't need to warn.
+        if method_name.name == sym::from_iter {
+            if let Some(trait_def_id) = self.tcx.lang_items().from_iterator_trait() {
+                if self.tcx.infer_ctxt().enter(|infcx| {
+                    let obligation = Obligation::new(
+                        ObligationCause::dummy(),
+                        ty::ParamEnv::empty(),
```

Comment:
This is almost certainly what's causing the crash. You need to get the ParamEnv from the function call I think? Not sure how this works for function bodies, I've only done it for types.

Reply:
See `enter_param_env` in src/librustdoc for how it works for types.

Reply:
Yes, definitely do not enter a new inference context here. 

Actually what would be really useful is if you wait just a bit for https://github.com/rust-lang/rust/pull/86901 to land, then you can call `self.infcx.type_implements_trait` which is pretty much exactly what you want here.

Reply:
For more context: each inference context defines the inference variables in scope and so forth. Since we already are in the middle of type checking, we have an active inference context, and the types and traits may reference variables defined in there. When you enter a new inference context, as you do here, you can't "import" types that may contain inference variables from some other context.

Reply:
Thanks for the explanation! That helps a lot. :)

Reply:
I didn't realize that `self.infcx` was available. Didn't look at the Deref implementation of FnCtxt ^^'.